lib/redis/distributed/partitioner.rb

```diff
@@ -0,0 +1,223 @@
+# frozen_string_literal: true
+
+require_relative '../errors'
+require_relative '../client'
+require_relative 'hash_ring'
+
+class Redis
+  class Distributed
+    # Partitioning with client side consistent hashing
+    class Partitioner
+      def initialize(options = {})
+        @options = options.dup
+        @tag_fmt = @options.dig(:distributed, :tag) || /^\{(.+?)\}/
+        @ring = @options.dig(:distributed, :ring) || HashRing.new
+        node_configs = @options.dig(:distributed, :nodes)
+        @options.delete(:distributed)
+        clients = build_node_clients(node_configs, @options)
+        clients.each { |c| @ring.add_node(c) }
+        @command = build_command_info(@ring.nodes)
+      end
+
+      def id
+        @ring.nodes.map(&:id).sort.join(', ')
+      end
+
+      def db
+        @ring.nodes.first.db
+      end
+
+      def db=(_db)
+        raise CannotDistribute, 'select'
```

Comment:
This error handling is on purpose because HashRing is broken by `SELECT` command. The checksum calculation depends on the [client id](https://github.com/redis/redis-rb/blob/v4.1.0/lib/redis/hash_ring.rb#L27).